acceptance-tests/src/test/java/it/com/atlassian/bitbucket/jenkins/internal/SmokeTest.java

```diff
@@ -651,11 +585,96 @@ private void runFullFlowMultiBranch(boolean triggerOnRefChange, boolean triggerO
 
         multiBranchJob.waitForBranchIndexingFinished(30);
 
-        //a trigger is bound to have triggered the build, so here we are satisfied to check that a new build was triggered
+        // We should now have a new build for the feature branch if branch discovery is enabled
+        verifyBuildStatus(multiBranchJob, featureBranchName, newFileCommitId, discoverBranches);
+
+        // We should now have a build for the PR if the PR webhook trigger and PR discovery have been enabled
+        verifyBuildStatus(multiBranchJob, prJobName, featureBranchCommitId,
+                discoverPullRequests && triggerOnPullRequest);
+    }
 
-        // Fetch and verify build status is posted for the new commit
-        fetchBuildStatusesFromBitbucket(newFileCommitId).forEach(status ->
-                assertThat(status, successfulBuildWithKey(featureBranchBuildName)));
+    private void runFullFlowMultiBranchWithManualReindexing(boolean discoverBranches, boolean discoverPullRequests)
+            throws IOException, GitAPIException {
+        BitbucketScmWorkflowMultiBranchJob multiBranchJob =
+                jenkins.jobs.create(BitbucketScmWorkflowMultiBranchJob.class);
+        BitbucketBranchSource bitbucketBranchSource = multiBranchJob.addBranchSource(BitbucketBranchSource.class);
+        bitbucketBranchSource
+                .credentialsId(bbsAdminCredsId)
+                .serverId(serverId)
+                .projectName(forkRepo.getProject().getKey())
+                .repositoryName(forkRepo.getSlug())
+                .discoverBranches(discoverBranches)
+                .discoverPullRequests(discoverPullRequests);
+        multiBranchJob.save();
+
+        // Clone (fork) repo
+        File checkoutDir = tempFolder.newFolder(REPOSITORY_CHECKOUT_DIR_NAME);
+        Git gitRepo = cloneRepo(ADMIN_CREDENTIALS_PROVIDER, checkoutDir, forkRepo);
+
+        // Push new Jenkinsfile to master
+        RevCommit masterCommit =
+                commitAndPushFile(gitRepo, ADMIN_CREDENTIALS_PROVIDER, MASTER_BRANCH_NAME, checkoutDir,
+                        JENKINS_FILE_NAME, ECHO_ONLY_JENKINS_FILE_CONTENT.getBytes(UTF_8));
+        String masterCommitId = masterCommit.getId().getName();
+
+        // Push Jenkinsfile to feature branch
+        final String branchName = "feature/test-feature";
+        gitRepo.branchCreate().setName(branchName).call();
+        RevCommit featureBranchCommit =
+                commitAndPushFile(gitRepo, ADMIN_CREDENTIALS_PROVIDER, branchName, checkoutDir, JENKINS_FILE_NAME,
+                        ECHO_ONLY_JENKINS_FILE_CONTENT.getBytes(UTF_8));
+        String featureBranchCommitId = featureBranchCommit.getId().getName();
+        triggerManualScan(multiBranchJob);
+
+        // Verify branch builds were triggered if branch discovery is enabled
+        verifyBuildStatus(multiBranchJob, MASTER_BRANCH_NAME, masterCommitId, discoverBranches);
```

Comment:
This feels very racy, you trigger a scan, and expect the build to be created and completed running in the next line of code, I know there is a `30s` wait in the `triggerManualScan` method but I worry that is not always enough.

Reply:
I could be misunderstanding it but the 30 seconds isn't just a wait but a timeout for a loop to wait for the indexing logs to print out a `Finished: `.
https://github.com/jenkinsci/acceptance-test-harness/blob/5504.v485694f31cdf/src/main/java/org/jenkinsci/test/acceptance/po/WorkflowMultiBranchJob.java#L43

Reply:
That waits for the _indexing_ to complete but we _also_ need the build that is found as part of the indexing to complete. 
I'm just looking at this with experience of flaky builds, and I see that `verifyBuildStatus` gets the build status of the last build and checks it. I am concerned that it will be flaky and work most of the time, but occasionally we will not have waited long enough for the correct build status to appear. 

I actually think you have a bug here as well. the `verifyBuildStatus` uses the last argument to indicate if the build should be successful or not, now we just pass in `discoverBranches` which.. is no longer related to the successful execution of the build.

This test has traditionally been very flaky so I'm trying to make sure that we improve it, not just maintain the status quo

Reply:
That makes sense. I'll see what I can do to improve these a bit. Thanks for your feedback @dkjellin!